src/jdk.jartool/share/classes/sun/tools/jar/Validator.java

```diff
@@ -62,20 +93,188 @@ final class Validator {
+     *
+     * NOTE: In order to check the encounter order based on the CEN listing,
+     *       this implementation assumes CEN entries are to be added before
+     *       add any LOC entries. That is, addCenEntry should be called before
+     *       calls to addLocEntry to ensure encounter order can be compared
+     *       properly.
+     */
+    private class EntryValidator {
+        // A place holder when an entry is not yet seen in the directory
+        static EntryEncounter PLACE_HOLDER = new EntryEncounter(0, 0);
+        // Flag to signal the CEN and LOC is not in the same order
+        boolean outOfOrder = false;
+        /**
+         * A record to keep the encounter order in the directory and count of the appearances
+         */
+        record EntryEncounter(int order, int count) {
+            /**
+             * Add to the appearance count.
+             * @param encounterOrder The supplier for the encounter order in the directory
+             */
+            EntryEncounter increase(IntSupplier encounterOrder) {
+                return isPlaceHolder() ?
+                    // First encounter of the entry in this directory
+                    new EntryEncounter(encounterOrder.getAsInt(), 1) :
+                    // After first encounter, keep the order but add the count
+                    new EntryEncounter(order, count + 1);
+            }
+
+            /**
+             * True if this entry is not in the directory.
+             */
+            boolean isPlaceHolder() {
+                return count == 0;
+            }
+        }
+
+        /**
+         * Information used for validation for a entry in CEN and LOC.
+         */
+        record EntryInfo(EntryEncounter cen, EntryEncounter loc) {}
+
+        /**
+         * Ordered deduplication set for entries
+         */
+        LinkedHashMap<String, EntryInfo> entries = new LinkedHashMap<>();
+        // Encounter order in CEN, step by 1 on each new entry
+        int cenEncounterOrder = 0;
+        // Encounter order in LOC, step by 1 for new LOC entry exist in CEN
+        int locEncounterOrder = 0;
+
+        /**
+         * Record an entry apperance in CEN
+         */
+        public void addCenEntry(ZipEntry cenEntry) {
+            var entryName = cenEntry.getName();
+            var entryInfo = entries.get(entryName);
+            if (entryInfo == null) {
+                entries.put(entryName, new EntryInfo(
+                        new EntryEncounter(cenEncounterOrder++, 1),
+                        PLACE_HOLDER));
+            } else {
+                assert entryInfo.loc().isPlaceHolder();
+                entries.put(entryName, new EntryInfo(
+                        entryInfo.cen().increase(() -> cenEncounterOrder++),
+                        entryInfo.loc()));
+            }
+        }
+
+        /**
+         * Record an entry apperance in LOC
+         * We compare entry order based on the CEN. Thus do not step LOC
```

Comment:
Nit - would "Thus do not increment LOC ..." be easier to understand? You don't have to do the change if you prefer it this way (I see that we use the same term on the comment of the `cenEncounterOrder` and `locEncounterOrder` fields.